src/php/NodeEnvironment.php

```diff
@@ -66,6 +71,15 @@ public static function empty(): NodeEnvironment
         return new NodeEnvironment([], self::CTX_STMT, [], []);
     }
 
+    public static function withDefNotAllowed(?NodeEnvironment $env = null): NodeEnvironment
```

Comment:
If you want such a method you should make it not static: `public function withDefNotAllowed()`. For the test you can easily build it like this: `NodeEnvironment::empty()->withDefNotAllowed()`.

I don't want to have too many static method for `NodeEnvironment` because they than may be used in the Analyzer in the wrong way. E.g. if you call `NodeEnvironment::withDefNotAllowed()`, this will create a new empty node environment and all other states are lost.